tools/i18n/unescape-contents

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python3
+
+import argparse
+import html
+import json
+import os
+import sys
+from contextlib import contextmanager
+from pathlib import PurePath
+from typing import IO, Any, Generator, NamedTuple, Union
+
+
+class CLIArgs(NamedTuple):
+    escape_html: bool
+    filename: str
+    indent_level: int
+
+
+def parse_args() -> CLIArgs:
+    parser = argparse.ArgumentParser(
+        prog="unescape-contents",
+        description="Unescape Unicode and, optionally, HTML entities in a JSON file. Input file must be a JSON dictionary. Output will always be unescaped UTF-8.",
+    )
+
+    parser.add_argument("filename", type=str)
+    parser.add_argument(
+        "--html",
+        action="store_true",
+        help="If the key of a dictionary field does not contain HTML escapes, unescape any HTML escapes found in the value",
+    )
+    parser.add_argument("--indent-level", dest="indent_level", type=int, default=2, required=False)
+    args = parser.parse_args()
+
+    return CLIArgs(args.html, args.filename, args.indent_level)
+
+
+@contextmanager
+def tempfile_for_atomic_overwrite(
+    destination: Union[str, "os.PathLike[str]"]
+) -> Generator[IO[Any], None, None]:
+    """
+    Creates a sibling temporary file in the same directory as `destination` and
+    yields the file-like object returned by `open()`. Upon completion of the
+    context block, the file handle is closed and the temporary file is removed,
+    either by replacing `destination` with its contents (see `os.replace()`
+    documentation) if the context succeeded, or by `os.unlink()`-ing it if the
+    context failed.
+
+    This is not a general-purpose temporary file creator: for usecases that
+    don't require moving the temporary file instead of deleting it, use the
+    `tempfile` module.
+    """
+
+    source_path = PurePath(os.path.abspath(destination))
+    temp_path = source_path.with_name(f".{source_path.name}.unescaped")
+    temp_file = open(temp_path, "w+")
+
+    try:
+        yield temp_file
+        temp_file.close()
+        os.replace(temp_file.name, destination)
+    finally:
+        try:
+            temp_file.close()
+            os.unlink(temp_path)
+        except FileNotFoundError:
+            # In the event of success, we've already moved this file
+            pass
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    print(f"unescaping file {args.filename}", file=sys.stderr)
+
+    with tempfile_for_atomic_overwrite(args.filename) as dest:
+        with open(args.filename) as source:
+            # WARNING: json.load will load the entire file into memory [1], and
+            # thus this should not be used on huge files. Our translation files
+            # are all <1MB in size, however, so in context this should be fine.
+            #
+            # [1] https://pythonspeed.com/articles/json-memory-streaming/
+            json_data = json.load(source)
```

Comment:
So I see the next block is where we unescape HTML; is it just doing JSON load/dump that escaping the JSON?

Reply:
Yeah - `json.dump` follows [these rules](https://docs.python.org/3/library/json.html#py-to-json-table) to output "as-is" via `ensure_ascii=False`. By observation, unicode escape sequences get converted into the actual Unicode glyphs at `load` time; [Python's docs seem to imply this might be locale specific](https://docs.python.org/3.10/howto/unicode.html#the-string-type) but works reliably in the Vagrant image and on my host Void Linux `en_US.UTF-8` system.

I have not robustly tested what happens on other OSes, other locales, etc.

Reply:
Added a comment in the source to this effect: this seems like a reasonable question any reader of this code could have and should have an inline answer to.